### DISCUSSION TITLE
How to edit keybindings on regolith sway?

### DISCUSSION THREAD

**Discussion by Fingel** [regolith-linux/regolith-desktop#919] [Q&A] [answered]
Hi,
I'm just trying out Regolith for the first time. It seems pretty great so far, except that I just want to change a few keybinding to those that I'm used to. For a keyboard driven desktop, you think this would be one of the first things anyone would do. But it does not seem simple?

The [configuration page](https://regolith-desktop.com/docs/using-regolith/configuration/#keybindings) has a section on keybindings, but only for the mod key (the default mod key is fine). It then links to another page for [Xresource settings](https://regolith-desktop.com/docs/reference/xresources/), but none of these are keybindings. Also, is Xresources even used for wayland/sway? '

Further digging brings you to the section in the docs for [configuring i3 settings](https://regolith-desktop.com/docs/howtos/customize-i3-configuration/) but again, not a single example for keybidings. 

Certainly, changing the default keybinding of something like "close application" should be easy to do? Or does regolith really want you to stick with the defaults?

Further reading on the i3 configuration section explains how to override partials of the i3 config locally, but the config files seem to reference `set_from_resource`, which again points to a Xresources file, but where to get the defaults? xrdb -query returns nothing, (probably because we are on wayland).

This seems insane.

**Comment by kgilmer** [comment, marked as answer]
Hi @Fingel !

Yes, the docs for this aspect of Regolith need some work.

> Also, is Xresources even used for wayland/sway?

There is an equivalent called trawl.  The client is `trawlcat`.  

> Certainly, changing the default keybinding of something like "close application" should be easy to do?

It should be yes.  Lets take your example for close application.  This is what I'd do:

Since I'm not sure where in the configuration the keybinding for "close application" is, I'd first search for it in Sway's partial dir:

```console
~/dev/repos/voulage$ find /usr/share/regolith/sway/config.d/ | xargs grep -HsFi exit
/usr/share/regolith/sway/config.d/55_session_keybindings:## Session // Exit App // <><Shift> q ##
/usr/share/regolith/sway/config.d/55_session_keybindings:set_from_resource $wm.binding.exit_app wm.binding.exit_app Shift+q
/usr/share/regolith/sway/config.d/55_session_keybindings:bindsym $mod+$wm.binding.exit_app [con_id="__focused__"] kill
/usr/share/regolith/sway/config.d/55_session_keybindings:set_from_resource $wm.program.logout wm.program.logout /usr/bin/gnome-session-quit --logout --no-prompt && swaymsg exit
```

It's not obvious but the line here that we care about is:

```
/usr/share/regolith/sway/config.d/55_session_keybindings:set_from_resource $wm.binding.exit_app wm.binding.exit_app Shift+q
```

So the keybinding defaults to `Shift+q` but we want to change that.  I'll try `Shift+x` for extra flair.  To override the value, you would add the line and your value to the Xresources file (yes it's confusing that in Sway we're dealing with Xresource files):

```
echo "wm.binding.exit_app: Shift+x" >> ~/.config/regolith3/Xresources
```

Then I refresh sway with `<super><shift>r` and am able to close apps via `<super><shift><x>`.

**Reply by Fingel** [reply]
This is great, thank you!

**Reply by worldofgeese** [reply]
Hi @kgilmer! I tried `find /usr/share/regolith/sway/config.d/ | xargs grep -HsFi screenshot` and `find /usr/share/regolith/sway/config.d/ | xargs grep -HsFi Print` but couldn't find any entry so created a partial: `cp /usr/share/regolith/sway/config.d/55_session_keybindings ~/.config/regolith3/sway/config.d/55_session_keybindings`. The [Regolith i3 docs on customizing](https://regolith-desktop.com/docs/howtos/customize-i3-configuration/#how-to-customize-a-default-partial) tell me to do the above, add any of my own customizations to the file then uninstall that specific partial package using `apt` but it's nowhere to be found for Sway.

 Here's my addition at the bottom of `~/.config/regolith3/sway/config.d/55_session_keybindings`:

`bindsym Print exec grim -g "$(slurp -d)" - | swappy -f -`. If I reload, the keybind doesn't work but the command does. What do I do?

**Reply by kgilmer** [reply]
@worldofgeese hmm, there may be an issue w/ sway interpreting the command.  You could tail `/var/log/syslog` after hitting the key and see if there is a specific error message.  There should be nothing custom to Regolith here, other than the location of the config partials.  So anything that works in a regular Sway setup should also work here.

**Reply by worldofgeese** [reply]
@kgilmer I figured this out thanks to tailing the logs as you suggested: `slurp`, `grim` and `swappy` weren't found. My Sway session isn't inheriting my `$PATH` correctly. I wonder if this is also the source of [my other issue I posted for illia](https://github.com/regolith-linux/ilia/issues/75)?

To fix, I needed to add the full path for each command: `bindsym Print exec /home/worldofgeese/.nix-profile/bin/grim -g "$(/home/worldofgeese/.nix-profile/bin/slurp -d)" - | /home/worldofgeese/.nix-profile/bin/swappy -f -`
